sklearn/metrics/_scorer.py

```diff
@@ -149,35 +157,47 @@ def __call__(self, estimator, *args, **kwargs):
                     scores[name] = format_exc()
         return scores
 
-    def _use_cache(self, estimator):
-        """Return True if using a cache is beneficial.
+    @staticmethod
+    def _reduce_response_method_scorer(scorer, estimator):
+        """To be cache friendly, we reduce the number of response methods to try for
+        a given estimator when `response_method` is a list or a tuple.
 
-        Caching may be beneficial when one of these conditions holds:
-          - `_ProbaScorer` will be called twice.
-          - `_PredictScorer` will be called twice.
-          - `_ThresholdScorer` will be called twice.
-          - `_ThresholdScorer` and `_PredictScorer` are called and
-             estimator is a regressor.
-          - `_ThresholdScorer` and `_ProbaScorer` are called and
-             estimator does not have a `decision_function` attribute.
+        We return a new scorer where the response method has been changed.
+        """
+        if isinstance(scorer._response_method, str):
+            return scorer
+        new_response_method = _check_response_method(
+            estimator, scorer._response_method
+        ).__name__
+
+        new_scorer = scorer.__class__(
+            scorer._score_func,
+            scorer._sign,
+            scorer._kwargs,
+            response_method=new_response_method,
+        )
+        return new_scorer
```

Comment:
what happens to metadata request values set on scorers here?

Reply:
It was a silent bug because we create internally a `scorers` from `self._scorers` that is used to make the routing.

Thus, forgetting to propagate the metadata routing will not impact the results because we were using `self._scorers`. I moved this part of the code in `_check_multimetric_scoring`: it avoids that `_MultimetricScorer` modifying is attributes and this silent bug and this is also the job of `_check_multimetric_scoring` to make this type of check.